### ISSUE TITLE
Issue with cuda.amp.autocast() during evaluation?

### ISSUE THREAD

**Issue by Aryan9101** [hushell/pmf_cvpr22#25]
I noticed that you guys have a torch.cuda.amp.autocast() in the _evaluate function ([line 143](https://github.com/hushell/pmf_cvpr22/blob/main/engine.py#L143)) in engine.py. In my experience, autocasting is typically used in conjunction with a GradScaler as well. Especially, given that said autocasting could be wrapping around the forward pass of a ProtoNet_Finetune or ProtoNet_Auto_Finetune model, which by itself, also includes multiple forward **_and_** backward passes (and PyTorch recommends not wrapping the latter with autocasting) through the backbone because of the final fine-tuning during the meta-testing stage.

 I ran two evaluations with the base DINO architecture (off-the-shelf and not meta-trained) on 100 5-way-5-shot classification tasks from the test split of CIFAR-FS, one with that [autocasting line](https://github.com/hushell/pmf_cvpr22/blob/main/engine.py#L143) commented out and with it as present in your codebase, and I noticed very different results. Here is the command I used by the way:
`python main.py --dataset cifar_fs --arch dino_small_patch16 --device cuda:0 --nSupport 5 --deploy finetune --eval --nEpisode 100 --ada_lr 0.001 --ada_steps 50`

When autocasting was commented out, the global average accuracy was around ~40.9% but with it present, it was around ~86.3%. 
<img width="1003" alt="Screenshot 2023-07-20 at 3 09 58 PM" src="https://github.com/hushell/pmf_cvpr22/assets/25390298/026d4e7c-0cc5-47e7-b484-35b5e616aa13">
<img width="997" alt="Screenshot 2023-07-20 at 3 19 10 PM" src="https://github.com/hushell/pmf_cvpr22/assets/25390298/f1060e15-e067-433d-ad49-91831ebfcc2e">

I also ran two evaluations where I let auto LR selection happen (as highlighted in your paper) instead of picking a learning rate myself, with this command:
`python main.py --dataset cifar_fs --arch dino_small_patch16 --device cuda:0 --nSupport 5 --deploy finetune_autolr --eval 
--nEpisode 100 --ada_steps 50`

With autocasting commented out, a non-zero learning rate was never picked so the final accuracy reported for each task was after protonet evaluation with the frozen backbone embeddings (this would essentially be equivalent to running the first command but with `ada_lr=0`) --- this yielded a final accuracy of ~78.6%. However, with autocasting present, as in line 143, non-zero learning rates did end up getting picked and the final accuracy was around ~84%.

I am just curious about this behavior. Admittedly, I am not too familiar with automatic mixed precision myself so I could totally be missing something here. Any thoughts on why including/removing torch.cuda.amp.autocast() could lead to such a drastic change?

**Comment by Aryan9101** [comment]
@hushell my apologies for the ping but I was hoping that one of the authors could comment on this? Again, I found your paper and codebase easy to understand but this was the only decision that stumped me!

**Comment by hushell** [comment]
@Aryan9101 I remember commenting out `autocast()` for [ProtoNet](https://github.com/hushell/pmf_cvpr22/blob/720b76d213cfda6c88113550c6dd40106b751ad2/models/protonet.py#L6) without fine-tuning works fine. Were you able to confirm this?
